Example notebooks for LoRA with custom models (#724)
huggingface/peft: docs/source/developer_guides/custom_models.mdx

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Working with custom models

> ## LoRA

Reviewer: I think we can remove this section header since there aren't any other fine-tuning techniques here.

Author: I think there will be more in the future, because other techniques such as IA³ should also work with custom models. If you prefer, I can still remove it now and then add it back in later.

Reviewer: Yeah, I think it'd be nicer if we structure it around the content we have right now, and it'd be easy to add it back in later when we have more content!

Author: Done

> Some fine-tuning techniques such as prompt tuning are specific to language models. That means that in PEFT, it is
> assumed that a 🤗 transformers is being used. Other fine-tuning techniques, however, are not restricted to specific
> model types. One such technique is [LoRA](https://arxiv.org/abs/2106.09685). In this guide, we will see how LoRA can be
> applied to non 🤗 transformers models.

Reviewer: ```suggestion
Some fine-tuning techniques, such as prompt tuning, are specific to language models. That means in 🤗 PEFT, it is assumed a 🤗 Transformers model is being used. However, other fine-tuning techniques - like [LoRA](./conceptual_guides/lora) - are not restricted to specific model types. 

In this guide, we will see how LoRA can be applied to a multilayer perception and a computer vision model from the [timm](https://huggingface.co/docs/timm/index) library.
```

> ### Example 1: Multilayer perceptron

Reviewer: ```suggestion
## Multilayer perceptron
```

Let's assume that we want to fine-tune a multilayer perceptron with LoRA. Here is the definition:

```python
from torch import nn

class MLP(nn.Module):
    def __init__(self, num_units_hidden=2000):
        super().__init__()
        self.seq = nn.Sequential(
            nn.Linear(20, num_units_hidden),
            nn.ReLU(),
            nn.Linear(num_units_hidden, num_units_hidden),
            nn.ReLU(),
            nn.Linear(num_units_hidden, 2),
            nn.LogSoftmax(dim=-1),
        )

    def forward(self, X):
        return self.seq(X)
```

This is a straightforward multilayer perceptron with an input layer, a hidden layer, and an output layer. 

> _Note: For this toy example, we choose an exceedingly large number of hidden units to highlight the efficiency gains
> thanks to PEFT, but those gains are in line with more realistic examples._

Reviewer: ```suggestion
<Tip>

For this toy example, we choose an exceedingly large number of hidden units to highlight the efficiency gains
from PEFT, but those gains are in line with more realistic examples.

</Tip>
```

> There are a few linear layers in this model that could be tuned with LoRA. When working with common 🤗 transformers

Reviewer: ```suggestion
There are a few linear layers in this model that could be tuned with LoRA. When working with common 🤗 Transformers
```

models, PEFT will know which layers to apply LoRA to, but in this case, it is up to us as a user to choose the layers.

> To achieve that, we first have to determine the names of the layers. The small snippet below prints them:

Reviewer: ```suggestion
To determine the names of the layers to tune:
```

 ('seq.2', torch.nn.modules.linear.Linear),
 ('seq.3', torch.nn.modules.activation.ReLU),
 ('seq.4', torch.nn.modules.linear.Linear),
 ('seq.5', torch.nn.modules.activation.LogSoftmax)]
```

Let's say we want to apply LoRA to the input layer and to the hidden layer, those are `'seq.0'` and `'seq.1'`. Moreover,
let's assume we want to update the output layer without LoRA, that would be `'seq.4'`. The corresponding config would
be:

```python
from peft import LoraConfig

config = LoraConfig(
    target_modules=["seq.0", "seq.2"],
    modules_to_save=["seq.4"],
)
```

With that, we can create our PEFT model and check the fraction of parameters trained:

```python
from peft import get_peft_model

model = MLP()
peft_model = get_peft_model(module, config)
peft_model.print_trainable_parameters()
# prints trainable params: 56,164 || all params: 4,100,164 || trainable%: 1.369798866581922
```

Finally, we can use any training framework we like, or write our own fit loop, to train the `peft_model`.

For a complete example, check out [this notebook](https://github.com/huggingface/peft/blob/main/examples/multilayer_perceptron/multilayer_perceptron_lora.ipynb).

> ### Example 2: A computer vision model from `timm`

Reviewer: ```suggestion
## timm model
```

 ('stages.0.blocks.0.mlp.act', torch.nn.modules.activation.GELU),
 ('stages.0.blocks.0.mlp.drop1', torch.nn.modules.dropout.Dropout),
 ('stages.0.blocks.0.mlp.norm', torch.nn.modules.linear.Identity),
 ('stages.0.blocks.0.mlp.fc2', torch.nn.modules.conv.Conv2d),
 ('stages.0.blocks.0.mlp.drop2', torch.nn.modules.dropout.Dropout),
 ('stages.0.blocks.0.drop_path2', torch.nn.modules.linear.Identity),
 ('stages.0.blocks.0.layer_scale2', timm.models.metaformer.Scale),
 ('stages.0.blocks.0.res_scale2', torch.nn.modules.linear.Identity),
 ('stages.0.blocks.1', timm.models.metaformer.MetaFormerBlock),
 ('stages.0.blocks.1.norm1', timm.layers.norm.GroupNorm1),
 ('stages.0.blocks.1.token_mixer', timm.models.metaformer.Pooling),
 ('stages.0.blocks.1.token_mixer.pool', torch.nn.modules.pooling.AvgPool2d),
 ...
 ('head.global_pool.flatten', torch.nn.modules.linear.Identity),
 ('head.norm', timm.layers.norm.LayerNorm2d),
 ('head.flatten', torch.nn.modules.flatten.Flatten),
 ('head.drop', torch.nn.modules.linear.Identity),
 ('head.fc', torch.nn.modules.linear.Linear)]
 ]
```

Upon closer inspection, we see that the 2D conv layers have names such as `"stages.0.blocks.0.mlp.fc1"` and

> `"stages.0.blocks.0.mlp.fc2"`. How can we match those layer names specifically? Thankfully, PEFT allows us to match them
> using [regular expressions](https://docs.python.org/3/library/re.html). For our case, the regex
> `r".*\.mlp\.fc\d"` should do the job.

Reviewer: ```suggestion
`"stages.0.blocks.0.mlp.fc2"`. How can we match those layer names specifically? You can write a [regular expressions](https://docs.python.org/3/library/re.html). to match the layer names. For our case, the regex
`r".*\.mlp\.fc\d"` should do the job.
```
